torch/csrc/jit/interpreter.cpp

```diff
@@ -19,6 +19,270 @@ namespace py = pybind11;
+//   should generate a move rather than a copy in this case.
+// * stage_input_types: the type annotations on the inputs to each stage
+//   these can be removed once the the backward tracer is no longer used
+
+struct PreprocessGraph {
+  PreprocessGraph(Graph & g)
+  : graph(g.copy()) {
+    desugarTripCounts(graph->block());
+    flattenStages();
+    dropUnused(graph->block());
+
+    // fill in move_flags by scanning blocks;
+    findLastUses(graph->block());
+    //TODO: desugar Loop trip counts, for now we drop trip counts
+  }
+  std::shared_ptr<Graph> graph;
+  // for each input, should we move rather than copy the inputs
+  std::unordered_map<Node*, std::vector<uint8_t>> move_flags;
+
+  // because JIT classic needs this to fix up gradients, remove when possible
+  std::vector<std::vector<TypePtr>> stage_input_types;
+private:
+
+  // this currently just _removes_ the trip count inputs and checks they are
+  // unused. In the future they will be desugared into normal arithmetic to
+  // provide a loop counter
+  void desugarTripCounts(Block * b) {
+    for(auto n : b->nodes()) {
+      if(n->kind() == kLoop) {
+
+        // remove the trip count from Loop inputs, we don't support it yet
+        n->removeInput(0);
+        JIT_ASSERT(n->blocks()[0]->inputs()[0]->uses().size() == 0 &&
+          "NYI - use of trip count variable");
+        JIT_ASSERT(n->blocks()[0]->inputs()[1]->uses().size() == 0 &&
+          "NYI - use of cond variable in loop");
+
+        // TODO: remove cond as input to loop carries, it just complicates this
+        // implementation
+        n->blocks()[0]->eraseInput(1);
+        n->blocks()[0]->eraseInput(0);
+      }
+      for(auto sb : n->blocks()) {
+        desugarTripCounts(sb);
+      }
+    }
+  }
+  Node * prependLocation() {
+    //handle corner case where there are no nodes,
+    if(graph->nodes().begin() == graph->nodes().end()) {
+      return graph->return_node();
+    }
+    return *graph->nodes().begin();
+  }
+  // removes all inputs and outputs to a graph, replacing them with nodes before of after each insertStage
+  void flattenStages() {
+    WithInsertPoint guard(*graph, prependLocation());
+    size_t input_pos = 0;
+    size_t output_pos = 0;
+    auto it = graph->nodes().begin();
+    for(size_t i = 0; i <= graph->stage(); i++) {
+      stage_input_types.emplace_back();
+      auto store = graph->insertNode(graph->create(kStore, 0));
+      // TODO: unused inputs need drop nodes added for them.
+      while(input_pos < graph->inputs().size() && graph->inputs()[input_pos]->stage() == i) {
+        auto nv = store->addOutput();
+        auto old_node = graph->inputs()[input_pos];
+        stage_input_types[i].push_back(old_node->typeOption());
+        old_node->replaceAllUsesWith(nv);
+        input_pos++;
+      }
+      while(it != graph->nodes().end() && it->stage() == i)
+        ++it;
+      graph->setInsertPoint(*it);
```

Comment:
Why not `WithInsertPoint`? This will change it permanently (until the next call), right?